In [60]:
import matplotlib.pyplot as plt 
import numpy as np 
import pandas as pd 
import seaborn as sns

df = pd.read_csv('/home/sary/cms-downloader/cms_downloads/Winter 2022/Data Engineering and Visualization/Project1/Car_sale_ads.csv')

In [61]:
df.head(10)

Index   Price Currency Condition Vehicle_brand Vehicle_model  \
0      0   86200      PLN       New        Abarth           595   
1      1   43500      PLN      Used        Abarth         Other   
2      2   44900      PLN      Used        Abarth           500   
3      3   39900      PLN      Used        Abarth           500   
4      4   97900      PLN       New        Abarth           595   
5      5   62950      PLN      Used        Abarth           595   
6      6   69900      PLN      Used        Abarth           595   
7      7   69900      PLN      Used        Abarth           500   
8      8  120750      PLN       New        Abarth           595   
9      9   93100      PLN       New        Abarth           595   

  Vehicle_version Vehicle_generation  Production_year  Mileage_km  ...  \
0             NaN                NaN             2021         1.0  ...   
1             NaN                NaN             1974     59000.0  ...   
2             NaN                NaN             2018     52000.0  ...   
3             NaN                NaN             2012     29000.0  ...   
4             NaN                NaN             2021       600.0  ...   
5             NaN                NaN             2016     46060.0  ...   
6             NaN                NaN             2021      2900.0  ...   
7             NaN                NaN             2019     20000.0  ...   
8             NaN                NaN             2021         1.0  ...   
9             NaN                NaN             2021         1.0  ...   

   Transmission        Type Doors_number  Colour Origin_country First_owner  \
0        Manual  small_cars          3.0    gray            NaN         NaN   
1        Manual       coupe          2.0  silver            NaN         NaN   
2     Automatic  small_cars          3.0  silver            NaN         NaN   
3        Manual  small_cars          3.0    gray            NaN         NaN   
4        Manual  small_cars          3.0    blue            NaN         NaN   
5        Manual  small_cars          3.0    gray        Germany         Yes   
6        Manual  small_cars          2.0   black        Germany         NaN   
7        Manual  small_cars          3.0   white         Poland         Yes   
8        Manual  small_cars          3.0    blue            NaN         NaN   
9        Manual  small_cars          3.0    gray            NaN         NaN   

  First_registration_date  Offer_publication_date  \
0                     NaN              04/05/2021   
1                     NaN              03/05/2021   
2                     NaN              03/05/2021   
3                     NaN              30/04/2021   
4                     NaN              30/04/2021   
5              15/03/2016              30/04/2021   
6              25/02/2021              29/04/2021   
7                     NaN              29/04/2021   
8                     NaN              03/05/2021   
9                     NaN              03/05/2021   

                                      Offer_location  \
0  ul. Jubilerska 6 - 04-190 Warszawa, Mazowiecki...   
1  kanonierska12 - 04-425 Warszawa, Rembertów (Po...   
2                   Warszawa, Mazowieckie, Białołęka   
3                                  Jaworzno, Śląskie   
4  ul. Gorzysława 9 - 61-057 Poznań, Nowe Miasto ...   
5  Modlińska 157 - 03-186 Warszawa, Białołęka (Po...   
6                             Żary, żarski, Lubuskie   
7     Ojcowska 2 - 02-918 Warszawa, Mokotów (Polska)   
8  ul. Gorzysława 9 - 61-057 Poznań, Nowe Miasto ...   
9  ul. Gorzysława 9 - 61-057 Poznań, Nowe Miasto ...   

                                            Features  
0                                                 []  
1                                                 []  
2  ['ABS', 'Electric front windows', 'Drivers air...  
3  ['ABS', 'Electric front windows', 'Drivers air...  
4  ['ABS', 'Electrically adjustable mirrors', 'Pa...  
5  ['ABS', 'Electric front windows', 'Drivers air

In [62]:
print(df.shape)
df.info()


(208304, 25)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208304 entries, 0 to 208303
Data columns (total 25 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    208304 non-null  int64  
 1   Price                    208304 non-null  int64  
 2   Currency                 208304 non-null  object 
 3   Condition                208304 non-null  object 
 4   Vehicle_brand            208304 non-null  object 
 5   Vehicle_model            208304 non-null  object 
 6   Vehicle_version          138082 non-null  object 
 7   Vehicle_generation       147860 non-null  object 
 8   Production_year          208304 non-null  int64  
 9   Mileage_km               207321 non-null  float64
 10  Power_HP                 207661 non-null  float64
 11  Displacement_cm3         206338 non-null  float64
 12  Fuel_type                208304 non-null  object 
 13  CO2_emissions            94047 non-null   floa

In [63]:
# null count
null_values = df.isnull().sum()
null_values

Index                           0
Price                           0
Currency                        0
Condition                       0
Vehicle_brand                   0
Vehicle_model                   0
Vehicle_version             70222
Vehicle_generation          60444
Production_year                 0
Mileage_km                    983
Power_HP                      643
Displacement_cm3             1966
Fuel_type                       0
CO2_emissions              114257
Drive                       15076
Transmission                  479
Type                            0
Doors_number                 1487
Colour                          0
Origin_country              89992
First_owner                143210
First_registration_date    121859
Offer_publication_date          0
Offer_location                  0
Features                        0
dtype: int64

In [64]:
# total number of cells
total_cells = np.product(df.shape)
total_missing = null_values.sum()
total_missing
print(f'total missing data: {total_missing/total_cells * 100}%')

total missing data: 11.917543590137491%


First owner option in the creation of the car ad is not required option(checkbox) so,
If you are not the first owner you will leave it which leads to null, and if you are then you will check it.
So replacing them with False will be suitable.

In [65]:

# replacing nulls in First_owner with False.
df['First_owner']=df['First_owner'].fillna(False)

#change yes in this column to True.
df['First_owner']=df['First_owner'].replace(['yes'],True)

#change the dtype of First_owner to bool.
df = df.astype({'First_owner':bool})

# drop First_registration_date as it contains > 58% missing data and can't be predicted from the other columns.
df.drop(columns=['First_registration_date'],inplace=True)

Some vehicle brands or models don't have version, so the only option is to choose: "I can't find my version" so we can replace it with "No version".
Also if you don't have version you will not have the generation.

Lets see the number of models that don't have version at all the rows.

Option#1 is to remove these columns as it contains 33% null values.
Option#2 is to replace nulls with "No version" as alternative value.


In [66]:
# Group the entries by the vehicle model (to count all the models we have)
groupByModel = df.groupby('Vehicle_model')


# remove the rows that contain nulls in the column Vehicle_version.
dfFilteredVersion = df[df['Vehicle_version'].notnull()]

# Use the filtered dataset to count number of models that have atleast one notnull value.
groupByAfterFilter = dfFilteredVersion.groupby('Vehicle_model')

print(groupByModel.sum().shape, groupByAfterFilter.sum().shape)

(1203, 9) (876, 9)


876 model that have Version but the seller just didn't enter it, and 327 model don't have any model.
So now we can remove this column as many sellers don't enter this value.(33% null values for Vehicle_version)

~30% null values at Vehicle_generation , it will be useless to drop Version and estimate the Generation as they are dependant.

In [67]:
total_rows = df.shape[0]
missing_Version = df[df['Vehicle_version'].isnull()].shape[0]
print(f'{missing_Version/total_rows * 100} % null values at Vehicle_version')

missing_Generation = df[df['Vehicle_generation'].isnull()].shape[0]
print(f'{missing_Generation/total_rows * 100} % null values at Vehicle_generation')

33.711306551962515 % null values at Vehicle_version
29.017205622551657 % null values at Vehicle_generation


In [68]:
# drop Vehicle_version column
df.drop(columns=['Vehicle_version'],inplace=True)

# drop Vehicle_generation column
df.drop(columns=['Vehicle_generation'], inplace=True)

CO2_emissions has ~55% null values and can't be predicted so removing this column will help.

In [69]:
# percentage of missing values at the column CO2 emissions
missing_CO2 = df[df['CO2_emissions'].isnull()].shape[0]
print(f'{missing_CO2/total_rows * 100} % null values at CO2_emissions')

54.851083032490976 % null values at CO2_emissions


In [70]:
# drop CO2_emissions
df.drop(columns=['CO2_emissions'],inplace=True)

Origin_country has ~43% null values and can't be predicted so removing this column will help.

In [71]:
# percentage of missing values at the column Origin_country
missing_Origin_country = df[df['Origin_country'].isnull()].shape[0]
print(f'{missing_Origin_country/total_rows * 100} % null values at Origin_country')

43.20224287579691 % null values at Origin_country


In [72]:
# drop Origin_country
df.drop(columns=['Origin_country'],inplace=True)

In [73]:
df.shape

(208304, 20)

In [74]:
null_values = df.isnull().sum()
null_values

Index                         0
Price                         0
Currency                      0
Condition                     0
Vehicle_brand                 0
Vehicle_model                 0
Production_year               0
Mileage_km                  983
Power_HP                    643
Displacement_cm3           1966
Fuel_type                     0
Drive                     15076
Transmission                479
Type                          0
Doors_number               1487
Colour                        0
First_owner                   0
Offer_publication_date        0
Offer_location                0
Features                      0
dtype: int64

Drive column is left optionally in the ad,but we can know it from the (car brand + vehicle model) columns.
and the rest of the rows which can't be estimated will replace it with "front_wheel".

In [75]:
df['Drive']

0         Front wheels
1         Front wheels
2                  NaN
3         Front wheels
4                  NaN
              ...     
208299             NaN
208300     Rear wheels
208301             NaN
208302             NaN
208303             NaN
Name: Drive, Length: 208304, dtype: object

In [79]:
# Save for each (car_brand + model) the drive of the car
df_filtered_drive = df.dropna(subset=['Drive'])
brand_modelToDrive = {}
for i in (df_filtered_drive['Index']):
    brand = df_filtered_drive.at[i,'Vehicle_brand']
    model = df_filtered_drive.at[i,'Vehicle_model']
    drive = df_filtered_drive.at[i,'Drive']
    brand_modelToDrive[brand+model] = drive


Now let's assign the drive that is not found with the corresponding brand+model

In [81]:
dfOnlyNulls = df.loc[df['Drive'].isnull()]
estimated_efficiently = 0
for i in dfOnlyNulls['Index']:
    brand = df.at[i, 'Vehicle_brand']
    model = df.at[i,'Vehicle_model']

    if (brand+model) in brand_modelToDrive.keys():
        df.at[i, 'Drive'] = brand_modelToDrive[brand+model]
        estimated_efficiently +=1
    else:
        df.at[i, 'Drive'] = 'Front wheels'
print(estimated_efficiently)

0


Wow we have 14941 Drive type prestored in our dataframe, and replaced correctly.
Only 135 rows will be replaced with Front wheel as we don't have drivetype for its brand+model. 

In [ ]:
df.isnull().sum()

Index                        0
Price                        0
Currency                     0
Condition                    0
Vehicle_brand                0
Vehicle_model                0
Production_year              0
Mileage_km                 983
Power_HP                   643
Displacement_cm3          1966
Fuel_type                    0
Drive                        0
Transmission               479
Type                         0
Doors_number              1487
Colour                       0
First_owner                  0
Offer_publication_date       0
Offer_location               0
Features                     0
dtype: int64

Observe that the currency is not only in PLN but 270 rows is in EUR.
We will convert the EUR to PLN.
1 EUR = 4.69 PLN at 1 December 2022


In [82]:
print(df['Currency'].value_counts())

PLN    208034
EUR       270
Name: Currency, dtype: int64


In [83]:
df_EUR = df.loc[df['Currency'] == 'EUR']
for i in df_EUR['Index']:
    df.at[i, 'Price'] = df.at[i, 'Price'] * 4.69

df_EUR = df.loc[df['Currency'] == 'EUR']
df_EUR

Index      Price Currency Condition Vehicle_brand Vehicle_model  \
141        141   86521.12      EUR       New         Aixam         Other   
142        142   83158.39      EUR       New         Aixam         Other   
143        143   66330.67      EUR       New         Aixam         Other   
144        144   62967.94      EUR       New         Aixam         Other   
152        152   45366.37      EUR       New         Aixam         Other   
...        ...        ...      ...       ...           ...           ...   
203703  203703  133665.00      EUR      Used         Volvo           S60   
208176  208176   88641.00      EUR      Used         Volvo         XC 90   
208205  208205   33768.00      EUR      Used         Volvo           V40   
208232  208232   41741.00      EUR      Used         Volvo         XC 60   
208281  208281   58625.00      EUR      Used      Wartburg         311.0   

        Production_year  Mileage_km  Power_HP  Displacement_cm3 Fuel_type  \
141                2021         1.0      12.0               NaN  Electric   
142                2021         1.0      12.0               NaN  Electric   
143                2021         1.0       8.0             479.0    Diesel   
144                2021         1.0       4.0             479.0    Diesel   
152                2021         1.0       8.0             400.0    Diesel   
...                 ...         ...       ...               ...       ...   
203703             2017     18000.0     367.0            1969.0  Gasoline   
208176             2015    115000.0     235.0            2000.0    Diesel   
208205             2017     54000.0     120.0            2000.0    Diesel   
208232             2016     79000.0     163.0            2000.0    Diesel   
208281             1960     96000.0      38.0             900.0  Gasoline   

                  Drive Transmission        Type  Doors_number Colour  \
141        Front wheels    Automatic  small_cars           2.0  other   
142        Front wheels    Automatic  small_cars           2.0  other   
143        Front wheels    Automatic  small_cars           2.0  other   
144        Front wheels    Automatic  small_cars           2.0  other   
152        Front wheels    Automatic  small_cars           3.0  white   
...                 ...          ...         ...           ...    ...   
203703     Front wheels    Automatic       sedan           4.0   blue   
208176  4x4 (permanent)    Automatic         SUV           5.0  other   
208205     Front wheels       Manual     compact           5.0  other   
208232     Front wheels    Automatic         SUV           5.0  other   
208281     Front wheels       Manual       sedan           4.0  other   

        First_owner Offer_publication_date  \
141           False             03/05/2021   
142           False             03/05/2021   
143           False             03/05/2021   
144           False             03/05/2021   
152           False             04/05/2021   
...             ...                    ...   
203703        False             03/05/2021   
208176         True             26/04/2021   
208205         True             26/04/2021   
208232         True             26/04/2021   
208281        False             28/04/2021   

                                           Offer_location  \
141     Borowska 258 - 50-558 Wrocław, Dolnośląskie (P...   
142     Borowska 258 - 50-558 Wrocław, Dolnośląskie (P...   
143     Borowska 258 - 50-558 Wrocław, Dolnośląskie (P...   
144     Borowska 258 - 50-558 Wrocław, Dolnośląskie (P...   
152     Borowska 258 - 50-558 Wrocław, Dolnośląskie (P...   
...                                                   ...   
203703                                 Radom, Mazowieckie   
208176      katowicka 68 - 43-100 Tychy, Śląskie (Polska)   
208205      katowicka 68 - 43-100 Tychy, Śląskie (Polska)   
208232      katowicka 68 - 43-100 Tychy, Śląskie (Polska)   
208281                     Pabianice, pabianicki, Łódzkie   

        